numpy/__init__.py

```diff
@@ -154,16 +157,28 @@
     from . import ma
     from . import matrixlib as _mat
     from .matrixlib import *
-    from .compat import long
-
-    # Make these accessible from numpy name-space
-    # but not imported in from numpy import *
-    # TODO[gh-6103]: Deprecate these
-    if sys.version_info[0] >= 3:
-        from builtins import bool, int, float, complex, object, str
-        unicode = str
-    else:
-        from __builtin__ import bool, int, float, complex, object, unicode, str
+
+    # Deprecations introduced in NumNy 1.18.0, 2019-11-11
+    import builtins as _builtins
+    __deprecated_attrs__.update({
+        n: (
+            getattr(_builtins, n),
+            "`np.{n}` is a deprecated alias for the builtin `{n}`. "
+            "Use `np.{n}_` if you meant the numpy type, and use `{n}` by "
+            "itself if you meant the builtin.".format(n=n)
+        )
+        for n in ["bool", "int", "float", "complex", "object", "str"]
+    })
+    __deprecated_attrs__.update({
+        n: (
+            getattr(compat, n),
+            "`np.{n}` is a deprecated alias for `np.compat.{n}`. "
+            "Use `np.{n}_` if you meant the numpy type, use `np.compat.{n}` "
+            "if you support python 2, and otherwise use `{n3}`."
+            .format(n=n, n3=n3)
+        )
+        for n, n3 in [("long", "int"), ("unicode", "str")]
+    })
```

Comment:
For reference, here's the original wording from @njsmith, which I'd forgotten about when I wrote this:
```
                "Writing 'np.{0}' is almost always a mistake. Historically "
                "and currently it is identical to writing plain '{0}' "
                "(i.e., it refers to the Python builtin), and at some point "
                "in the future it will become an error. Replace with '{0}', "
                "or if you want the numpy-specific type, use 'np.{1}'."
                .format(_name, _numpy_equiv)
```

If anyone prefers parts of this wording, I'd appreciate if they could suggest a combined version as a reply to this comment.

Reply:
The majority of uses is likely as a `dtype=` argument, in which case the differences do not really matter (unless we change `int` on windows).  I do not prefer the history/mistake information though, its cute for super-users but probably confusing for most users looking to update code quickly.

I am considering appending something like ```"Most users should prefer the Python builtins `{0}(2)` when creating scalars, and `dtype=np.{0}_` when creating arrays.```?

But, I admit e.g. for `bool` and `object` it does not really matter and the Python scalars are just as good to signal `dtype`.

Reply:
I think we maybe should just use the same message in the second case with Py2 gone, except that it needs the two different names as inputs.

Reply:
Can you elaborate on those comments with a suggested change?

Reply:
Sorry, a bit nitpicking and brainstorming... We are going to spend some "annoying users" chips on this, so wanted to get this right.

But, I have asked (non super-user), and rethought... And I think we should just keep it as is.  I am not sure the typical user will really understand it. But the main point is: The typical user will not care at all, they will simply choose either of the replacements and be happy that it works, assuming (correctly) that if it likely mattered, we would tell them.

If we start getting questions We could add a URL to keep the message short while providing details elsewhere.

Reply:
Perhaps the message could emphasize that the way to make the warning go away without changing anything is to use `int` instead of `np.int`. The current warning steers them more towards `int_`, which is marginally more likely to break something.

Reply:
I like that idea. As mentioned above, I think I slightly prefer `float64` over `float_` as well. But no need to squabble about it.

Reply:
The reason for mentioning `float_` is that people will then hopefully spot the rule that the numpy types are the builtins with a `_` appended.

Reply:
True, sounds fine to.  I am happy with whatever small (or none) you still want to make, I don't have any more concrete proposals that may make a difference anymore.

If anyone else comes up with concrete proposals that is fine of course.

The DeprecationTest would be the main thing I would want to add before merging, but its not there in a few days, I may just add it myself (pinged the mailing list just in case, so want to wait a tiny bit).

Reply:
Updated with the test.